lib/runtime/formatter.js

```diff
@@ -0,0 +1,43 @@
+/** @babel */
+
+import { client } from '../connection'
+
+const { format } = client.import(['format'])
+
+
+// @TODO: Once DocumentFormat.jl comes to be able to handle options to specify indents,
+//        we can extract some information from the status-bar tile consumed by atom-indent-detective,
+//        or more safely, make atom-indent-detective provide the service and use it here ?
+
+export function formatCode () {
+  const editor = atom.workspace.getActiveTextEditor()
+  if (!editor) return
+
+  const selections = editor.getSelections()
+  if (selections.length === 1 && !selections[0].getText()) {
+    formatEditor(editor)
+  } else {
+    selections.forEach((selection) => {
+      formatSelection(editor, selection)
+    })
+  }
+}
+
+function formatEditor (editor) {
+  format({
+    text: editor.getText(),
+    // indent: INDENT // @TODO
+  }).then(({ formattedtext }) => {
+    editor.setText(formattedtext)
+  })
+}
+
+function formatSelection(editor, selection) {
+  const range = selection.getBufferRange()
+  format({
+    text: selection.getText(),
+    // indent: INDENT // @TODO
+  }).then(({ formattedtext }) => {
+    editor.setTextInBufferRange(range, formattedtext)
```

Comment:
What happens if I add a new line somewhere else before the `Promise` is fulfilled?
It might make sense to use the marker API here (and we can also add a decoration to show what we're formatting). That'd make it possible to 
- keep track of the selected text no matter what other edits were made to the rest of the document
- don't apply the formatting if the content of the marked range changes

Reply:
The current implementation just _overwrites_ the text inside of the range even when it's modified before the `Promise` gets fullfilled.

And I think the marker would be more friendly and nicer, thus I will give a try to it.

Maybe is there some utility for marking a editor like the one we use on code execution ?

Reply:
I wondered `ink.highlight` can be useful here, but it always destroy markers after 20ms now, thus I will implement creating newer function in `formatter.js`

Reply:
The decoration stuff isn't as important as tracking the right selection and canceling the formatting if the user makes changes, so don't worry about it too much.

Reply:
I see. I've done the work in the last commit 😄